README.md

<p align="center">
<img src="https://github.com/homebridge/branding/raw/master/logos/homebridge-wordmark-logo-vertical.png" width="150" alt="homebridge">
</p>
<p align="center">
<a href="https://github.com/homebridge/homebridge/wiki/Verified-Plugins"><img src="https://badgen.net/badge/homebridge/verified/purple" alt="verified-by-homebridge"></a>
</p>


# Homebridge Nuptun Smart

This plugin provides an integration with Neptun Smart leak prevention system.
No online API. Control your sensors and valves via local network using ModBus RTU.

## Instructions

1. Connect your Neptun Smart appliance to Wi-Fi network.
2. Discover device IP address and make static DHCP lease settings.
3. Install plugin and configure it using [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x). 
4. Minimum required settings are: `id`, `displayName`, `host`, `groupsEnabled`, `wiredSensorsCount`.

### Wireless sensors will be discovered automatically.

## Example Config

If you are not using [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x) feel free to use this example config.

```json
{
  "platforms": [
    {
      "devices" : [
        {
          "id" : "neptun-1",
          "displayName" : "Neptun Smart Bath",
          "host" : "192.168.1.30",
          "groupsEnabled" : false,
          "wiredSensorsCount" : 3
        },
        {
          "id" : "neptun-2",
          "displayName" : "Neptun Smart Shower",
          "host" : "192.168.1.31",
          "port" : 503,
          "address": 240,
          "groupsEnabled" : false,
          "wiredSensorsCount" : 4,
          "updateInterval" : 10,
        }
      ],
      "platform" : "NeptunSmart"
    }
  ]
}
```

CHANGELOG.md
<a name="v1.0.10"></a>
# [v1.0.10](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.10) - 11 May 2022

**fix:** Now using unique name for valves of the second group

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.9...v1.0.10

[Changes][v1.0.10]


<a name="v1.0.9"></a>
# [v1.0.9](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.9) - 28 Apr 2022

**fix:** Characteristics were changing for wrong valve
**fix:** Now using `Valves` instead of `Faucets`

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.7...v1.0.9

[Changes][v1.0.9]


<a name="v1.0.7"></a>
# [v1.0.7](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.7) - 27 Apr 2022

**fix:** Wrong array was used for wireless sensors

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.6...v1.0.7

[Changes][v1.0.7]


<a name="v1.0.6"></a>
# [v1.0.6](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.6) - 27 Apr 2022

**fix:** No continuous pushing to array of wireless sensors

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.5...v1.0.6

[Changes][v1.0.6]


<a name="v1.0.5"></a>
# [v1.0.5](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.5) - 27 Apr 2022

**fix:** Both valves were acting the same when grouping is enabled

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.4...v1.0.5

[Changes][v1.0.5]


<a name="v1.0.4"></a>
# [v1.0.4](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.4) - 27 Apr 2022

**feature:** Instructions in the README
**fix:** Switching both valves when grouping is disabled

**Full Changelog**: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.0...v1.0.4

[Changes][v1.0.4]


<a name="v1.0.0"></a>
# [v1.0.0](https://github.com/maxim-smirnov/homebridge-neptun-smart/releases/tag/v1.0.0) - 27 Apr 2022

Initial public release.

[Changes][v1.0.0]


[v1.0.10]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.9...v1.0.10
[v1.0.9]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.7...v1.0.9
[v1.0.7]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.6...v1.0.7
[v1.0.6]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.5...v1.0.6
[v1.0.5]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.4...v1.0.5
[v1.0.4]: https://github.com/maxim-smirnov/homebridge-neptun-smart/compare/v1.0.0...v1.0.4
[v1.0.0]: https://github.com/maxim-smirnov/homebridge-neptun-smart/tree/v1.0.0

 <!-- Generated by https://github.com/rhysd/changelog-from-release -->


src/platform.ts
import { API, DynamicPlatformPlugin, Logger, PlatformAccessory, PlatformConfig, Service, Characteristic } from 'homebridge';

import { PLATFORM_NAME, PLUGIN_NAME } from './settings';
import { NeptunSmart } from './neptunSmart';

/**
 * HomebridgePlatform
 * This class is the main constructor for your plugin, this is where you should
 * parse the user config and discover/register accessories with Homebridge.
 */
export class NeptunSmartPlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  // this is used to track restored cached accessories
  public readonly accessories: PlatformAccessory[] = [];

  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig,
    public readonly api: API,
  ) {
    // When this event is fired it means Homebridge has restored all cached accessories from disk.
    // Dynamic Platform plugins should only register new accessories after this event was fired,
    // in order to ensure they weren't added to homebridge already. This event can also be used
    // to start discovery of new accessories.
    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback');
      for (const accessory of this.accessories) {
        if (!this.config.devices.map(device => this.api.hap.uuid.generate(device.id)).includes(accessory.UUID)) {
          // accessory removed
          log.info('Removed cached accessory: ' + accessory.displayName);
          this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
        }
      }
      // run the method to discover / register your devices as accessories
      this.discoverDevices();
    });
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);
    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.accessories.push(accessory);
  }

  /**
   * This is an example method showing how to register discovered accessories.
   * Accessories must only be registered once, previously created accessories
   * must not be registered again to prevent "duplicate UUID" errors.
   */
  discoverDevices() {

    // loop over the discovered devices and register each one if it has not already been registered
    for (const device of this.config.devices) {

      // generate a unique id for the accessory this should be generated from
      // something globally unique, but constant, for example, the device serial
      // number or MAC address
      const uuid = this.api.hap.uuid.generate(device.id);

      // see if an accessory with the same uuid has already been registered and restored from
      // the cached devices we stored in the `configureAccessory` method above
      const existingAccessory = this.accessories.find(accessory => accessory.UUID === uuid);

      if (existingAccessory) {
        // the accessory already exists
        this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);

        // if you need to update the accessory.context then you should run `api.updatePlatformAccessories`. eg.:
        // existingAccessory.context.device = device;
        // this.api.updatePlatformAccessories([existingAccessory]);

        // create the accessory handler for the restored accessory
        // this is imported from `neptunSmart.ts`
        new NeptunSmart(this, existingAccessory, device);

        // it is possible to remove platform accessories at any time using `api.unregisterPlatformAccessories`, eg.:
        // remove platform accessories when no longer present
        // this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [existingAccessory]);
        // this.log.info('Removing existing accessory from cache:', existingAccessory.displayName);
      } else {
        // the accessory does not yet exist, so we need to create it
        this.log.info('Adding new accessory:', device.id);

        // create a new accessory
        const accessory = new this.api.platformAccessory(device.displayName, uuid);

        // store a copy of the device object in the `accessory.context`
        // the `context` property can be used to store any data about the accessory you may need
        accessory.context.device = device;

        // create the accessory handler for the newly create accessory
        // this is imported from `neptunSmart.ts`
        new NeptunSmart(this, accessory, device);

        // link the accessory to your platform
        this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
      }
    }
  }
}


src/neptunSmartData.ts
export enum Register {
    // 13-bit value
    Config = 0,
    // 4-bit value
    WiredSensorsStatus = 3,
    // Number of wireless sensors connected
    WirelessSensorsCount = 6,
    // First wireless sensor register
    WirelessSenorStatus = 57
}

// Neptun Smart device config
export interface Config {
    floorWashMode: boolean; // 0
    alarmFirstGroup: boolean; // 1
    alarmSecondGroup: boolean; // 2
    wirelessSensorsLowBattery: boolean; // 3
    wirelessSensorsConnectionLost: boolean; // 4
    valveCloseOnFirstGroupLostConnection: boolean; // 5
    valveCloseOnSecondGroupLostConnection: boolean; // 6
    paringModeActive: boolean; // 7
    valveOpenFirstGroup: boolean; // 8
    valveOpenSecondGroup: boolean; // 9
    groupsEnabled: boolean; // A
    valveCloseOnLostConnection: boolean; // B
    keyboardLocked: boolean; // C
}

// Wired sensors status
export interface WiredSensorsStatus {
    alarmDetectedLineOne: boolean; // 0
    alarmDetectedLineTwo: boolean; // 1
    alarmDetectedLineThree: boolean; // 2
    alarmDetectedLineFour: boolean; // 3
}

// Number of wireless sensors connected
export interface WirelessSensorsCount {
    wirelessSensorsCount: number;
}

// Structure of each wireless sensor status
export interface WirelessSenorStatus {
    alarmDetected: boolean; // 0
    connectionLost: boolean; // 2
    batteryLevel: number; // F-8
}

src/neptunSmartModbus.ts
import ModbusRTU from 'modbus-serial';
import {throttle} from 'throttle-debounce';
import {ReadRegisterResult, WriteRegisterResult} from 'modbus-serial/ModbusRTU';
import {Config, Register, WiredSensorsStatus, WirelessSenorStatus, WirelessSensorsCount} from './neptunSmartData';

export declare type ConfigWriter = (value: Config) => Promise<Config>;

export class NeptunSmartModbus {

  private client: ModbusRTU | null = null;

  private networkErrors = [
    'ESOCKETTIMEDOUT',
    'ETIMEDOUT',
    'ECONNRESET',
    'ECONNREFUSED',
    'EHOSTUNREACH',
    'ENETRESET',
    'ECONNABORTED',
    'ENETUNREACH',
    'ENOTCONN',
    'ESHUTDOWN',
    'EHOSTDOWN',
    'ENETDOWN',
    'EWOULDBLOCK',
    'EAGAIN',
  ];

  constructor(
    private readonly host: string,
    private readonly port: number,
    private readonly address: number,
    private readonly didConnect: () => void,
  ) {
    this.connect();
  }

  private connect() {
    this.client = null;

    const client = new ModbusRTU();
    client.connectTCP(this.host, { port: this.port })
      .then(() => {
        client.setID(this.address);
        client.setTimeout(5000);
        this.client = client;
        this.didConnect();
      })
      .catch((e) => {
        this.checkError(e, true);
      });
  }

  private throttledReconnect = throttle(10000, () => {
    if (this.client !== null && this.client.isOpen) {
      this.client.close(() => {
        this.connect();
      });
    } else {
      this.connect();
    }
  });

  // eslint-disable-next-line
  private checkError(e: any, forceRetry: boolean = false) {
    if((e.message && this.networkErrors.includes(e.message))
      || (e.errno && this.networkErrors.includes(e.errno))
      || forceRetry) {
      this.throttledReconnect();
    }
  }

  private getBit = (value: number, bit: number): number => {
    return (value & (1 << bit)) >> bit;
  };

  async fetchConfig(): Promise<Config> {
    const configNumber = await this.readSingleRegister(Register.Config);
    const config : Config = {
      floorWashMode: !!this.getBit(configNumber, 0), // 0
      alarmFirstGroup: !!this.getBit(configNumber, 1), // 1
      alarmSecondGroup: !!this.getBit(configNumber, 2), // 2
      wirelessSensorsLowBattery: !!this.getBit(configNumber, 3), // 3
      wirelessSensorsConnectionLost: !!this.getBit(configNumber, 4), // 4
      valveCloseOnFirstGroupLostConnection: !!this.getBit(configNumber, 5), // 5
      valveCloseOnSecondGroupLostConnection: !!this.getBit(configNumber, 6), // 6
      paringModeActive: !!this.getBit(configNumber, 7), // 7
      valveOpenFirstGroup: !!this.getBit(configNumber, 8), // 8
      valveOpenSecondGroup: !!this.getBit(configNumber, 9), // 9
      groupsEnabled: !!this.getBit(configNumber, 10), // A
      valveCloseOnLostConnection: !!this.getBit(configNumber, 11), // B
      keyboardLocked: !!this.getBit(configNumber, 12), // C
    };

    return config;
  }

  async fetchWiredSensorsStatus(): Promise<WiredSensorsStatus> {
    const wiredSensorsStatusNumber = await this.readSingleRegister(Register.WiredSensorsStatus);
    const wiredSensorsStatus: WiredSensorsStatus = {
      alarmDetectedLineOne: !!this.getBit(wiredSensorsStatusNumber, 0),
      alarmDetectedLineTwo: !!this.getBit(wiredSensorsStatusNumber, 1),
      alarmDetectedLineThree: !!this.getBit(wiredSensorsStatusNumber, 2),
      alarmDetectedLineFour: !!this.getBit(wiredSensorsStatusNumber, 3),
    };

    return wiredSensorsStatus;
  }

  async fetchWirelessSensorsCount(): Promise<WirelessSensorsCount> {
    const wirelessSensorsCount: WirelessSensorsCount = {
      wirelessSensorsCount: await this.readSingleRegister(Register.WirelessSensorsCount),
    };

    return wirelessSensorsCount;
  }

  async fetchWirelessSensorStatus(i: number): Promise<WirelessSenorStatus> {
    const wirelessSenorStatusNumber = await this.readSingleRegister(Register.WirelessSenorStatus + i);
    const wirelessSensorStatus: WirelessSenorStatus = {
      alarmDetected: !!this.getBit(wirelessSenorStatusNumber, 0),
      connectionLost: !!this.getBit(wirelessSenorStatusNumber, 2),
      batteryLevel: wirelessSenorStatusNumber >> 8,
    };
    return wirelessSensorStatus;
  }

  private async readSingleRegister(register: Register): Promise<number> {
    return this.readHoldingRegisters(register, 1)
      .then((result) => {
        if (result.data.length === 0) {
          throw Error('No result returned.');
        }
        return result.data[0];
      });
  }

  private async readHoldingRegisters(dataAddress: number, length: number): Promise<ReadRegisterResult> {
    if (this.client === null) {
      return Promise.reject(new Error('Disconnected.'));
    }
    return this.client.readHoldingRegisters(dataAddress, length)
      .catch((e) => {
        this.checkError(e);
        throw e;
      });
  }

  public async writeConfigRegister(value: Config): Promise<Config> {
    const modbusValue = +value.floorWashMode +
      (+value.alarmFirstGroup << 1) +
      (+value.alarmSecondGroup << 2) +
      (+value.wirelessSensorsLowBattery << 3) +
      (+value.wirelessSensorsConnectionLost << 4) +
      (+value.valveCloseOnFirstGroupLostConnection << 5) +
      (+value.valveCloseOnSecondGroupLostConnection << 6) +
      (+value.paringModeActive << 7) +
      (+value.valveOpenFirstGroup << 8) +
      (+value.valveOpenSecondGroup << 9) +
      (+value.groupsEnabled << 10) +
      (+value.valveCloseOnLostConnection << 11) +
      (+value.keyboardLocked << 12);
    const configNumber = await this.writeSingleRegister(Register.Config, modbusValue);
    const config : Config = {
      floorWashMode: !!this.getBit(configNumber, 0), // 0
      alarmFirstGroup: !!this.getBit(configNumber, 1), // 1
      alarmSecondGroup: !!this.getBit(configNumber, 2), // 2
      wirelessSensorsLowBattery: !!this.getBit(configNumber, 3), // 3
      wirelessSensorsConnectionLost: !!this.getBit(configNumber, 4), // 4
      valveCloseOnFirstGroupLostConnection: !!this.getBit(configNumber, 5), // 5
      valveCloseOnSecondGroupLostConnection: !!this.getBit(configNumber, 6), // 6
      paringModeActive: !!this.getBit(configNumber, 7), // 7
      valveOpenFirstGroup: !!this.getBit(configNumber, 8), // 8
      valveOpenSecondGroup: !!this.getBit(configNumber, 9), // 9
      groupsEnabled: !!this.getBit(configNumber, 10), // A
      valveCloseOnLostConnection: !!this.getBit(configNumber, 11), // B
      keyboardLocked: !!this.getBit(configNumber, 12), // C
    };
    return config;
  }

  private async writeSingleRegister(register: Register, value: number): Promise<number> {
    return this.writeRegister(register, value)
      .then((result) => {
        if (result.value !== value) {
          throw Error('Setting value failed.');
        }
        return result.value;
      });
  }

  private async writeRegister(dataAddress: number, value: number): Promise<WriteRegisterResult> {
    if (this.client === null) {
      return Promise.reject(new Error('Disconnected.'));
    }
    return this.client.writeRegister(dataAddress, value)
      .catch((e) => {
        this.checkError(e);
        throw e;
      });
  }
}

src/neptunSmart.ts
import { PlatformAccessory, CharacteristicSetCallback } from 'homebridge';

import { NeptunSmartPlatform } from './platform';
import {NeptunSmartModbus, ConfigWriter} from './neptunSmartModbus';
import {Config} from './neptunSmartData';

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
export class NeptunSmart {
  private valvesServices;
  private wiredSensorServices;
  private wirelessSensorServices;
  readonly id;
  readonly displayName;
  private readonly host;
  private readonly port;
  private readonly address;
  private readonly updateInterval;
  private readonly wiredSensorsCount;
  private readonly isGroupsEnabled;
  private currentConfig;

  private neptunSmartModbus: NeptunSmartModbus;

  constructor(
    private readonly platform: NeptunSmartPlatform,
    private readonly accessory: PlatformAccessory,
    private readonly config,
  ) {
    // load all information from context
    this.id = accessory.context.device.id;
    this.displayName = accessory.context.device.displayName;
    this.host = config.host;
    this.port = config.port ?? 503;
    this.address = config.address ?? 240;
    this.wiredSensorsCount = config.wiredSensorsCount;
    this.updateInterval = config.updateInterval ?? 60;
    this.isGroupsEnabled = config.groupsEnabled ?? false;

    this.valvesServices = [];
    this.wiredSensorServices = [];
    this.wirelessSensorServices = [];

    this.platform.log.debug(this.wiredSensorsCount);
    this.neptunSmartModbus = new NeptunSmartModbus(this.host, this.port, this.address, () => {
      this.setUpAfterConnection();
    });

    // set accessory information
    this.accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Neptun')
      .setCharacteristic(this.platform.Characteristic.Model, 'Neptun Smart');

    this.valvesServices.push(this.accessory.getService('Valves group 1') ||
        this.accessory.addService(this.platform.Service.Valve, 'Valves group 1', 'neptun-valves-group-1'));
    this.valvesServices[0].setCharacteristic(this.platform.Characteristic.Name, 'Valves group 1');
    this.valvesServices[0].setCharacteristic(
      this.platform.Characteristic.ValveType, this.platform.Characteristic.ValveType.GENERIC_VALVE);
    this.valvesServices[0].setCharacteristic(this.platform.Characteristic.InUse, this.platform.Characteristic.InUse.IN_USE);
    this.valvesServices[0].getCharacteristic(this.platform.Characteristic.Active)
      .onSet(this.handleFaucetFirstGroupActiveSet.bind(this));
    if (this.isGroupsEnabled) {
      this.valvesServices.push(this.accessory.getService('Valves group 2') ||
          this.accessory.addService(this.platform.Service.Valve, 'Valves group 2', 'neptun-valves-group-2'));
      this.valvesServices[1].setCharacteristic(this.platform.Characteristic.Name, 'Valves group 2');
      this.valvesServices[1].setCharacteristic(
        this.platform.Characteristic.ValveType, this.platform.Characteristic.ValveType.GENERIC_VALVE);
      this.valvesServices[1].setCharacteristic(this.platform.Characteristic.InUse, this.platform.Characteristic.InUse.IN_USE);
      this.valvesServices[1].getCharacteristic(this.platform.Characteristic.Active)
        .onSet(this.handleFaucetSecondGroupActiveSet.bind(this));
    }

    for (let i=0; i<this.wiredSensorsCount; ++i) {
      this.wiredSensorServices.push(this.accessory.getService(`Wired sensor ${i+1}`) ||
          this.accessory.addService(this.platform.Service.LeakSensor, `Wired sensor ${i+1}`, `neptun-wired-sensor-${i+1}`));
      this.wiredSensorServices[i].setCharacteristic(this.platform.Characteristic.Name, `Wired sensor ${i+1}`);
    }

    setInterval(() => {
      this.updateFromDevice(platform);
    }, this.updateInterval * 1000);
  }

  private async setUpAfterConnection() {
    try {
      const config = await this.neptunSmartModbus.fetchConfig();
      this.platform.log.debug('Updating config:', config);

      this.accessory.getService(this.platform.Service.AccessoryInformation)!
        .setCharacteristic(this.platform.Characteristic.SerialNumber, this.id);

      this.platform.log.info('Device with id', this.id, 'discovered successfully');

    } catch (e) {
      if(e instanceof Error) {
        this.platform.log.error('Could obtain device metadata after connection.', e.message);
      }
    }
  }

  private async updateFromDevice(platform: NeptunSmartPlatform) {
    try {
      const config = this.currentConfig = await this.neptunSmartModbus.fetchConfig();
      this.platform.log.debug('Updating with config:', config);

      const c = platform.Characteristic;

      this.valvesServices[0].updateCharacteristic(c.Active, config.valveOpenFirstGroup ? c.Active.ACTIVE : c.Active.INACTIVE);
      if (this.isGroupsEnabled) {
        this.valvesServices[1].updateCharacteristic(c.Active, config.valveOpenSecondGroup ? c.Active.ACTIVE : c.Active.INACTIVE);
      }

      const wiredSensorsStatus = await this.neptunSmartModbus.fetchWiredSensorsStatus();
      this.platform.log.debug('Updating with wiredSensorsStatus:', wiredSensorsStatus);

      if (this.wiredSensorsCount > 0) {
        if (wiredSensorsStatus.alarmDetectedLineOne) {
          this.wiredSensorServices[0].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_DETECTED);
        } else {
          this.wiredSensorServices[0].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_NOT_DETECTED);
        }
      }

      if (this.wiredSensorsCount > 1) {
        if (wiredSensorsStatus.alarmDetectedLineTwo) {
          this.wiredSensorServices[1].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_DETECTED);
        } else {
          this.wiredSensorServices[1].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_NOT_DETECTED);
        }
      }

      if (this.wiredSensorsCount > 2) {
        if (wiredSensorsStatus.alarmDetectedLineThree) {
          this.wiredSensorServices[2].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_DETECTED);
        } else {
          this.wiredSensorServices[2].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_NOT_DETECTED);
        }
      }

      if (this.wiredSensorsCount > 3) {
        if (wiredSensorsStatus.alarmDetectedLineFour) {
          this.wiredSensorServices[3].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_DETECTED);
        } else {
          this.wiredSensorServices[3].updateCharacteristic(c.LeakDetected, c.LeakDetected.LEAK_NOT_DETECTED);
        }
      }

      const wirelessSensorsCount = await this.neptunSmartModbus.fetchWirelessSensorsCount();
      this.platform.log.debug('Updating with wirelessSensorsCount:', wirelessSensorsCount);

      if (this.wirelessSensorServices.length !== wirelessSensorsCount.wirelessSensorsCount) {
        this.wirelessSensorServices.map(service => this.accessory.removeService(service));
        this.wirelessSensorServices.slice(0, this.wirelessSensorServices.length);
        for (let i=0; i<wirelessSensorsCount.wirelessSensorsCount; ++i) {
          this.wirelessSensorServices.push(this.accessory.getService(`Wireless sensor ${i + 1}`) ||
            this.accessory.addService(this.platform.Service.LeakSensor, `Wireless sensor ${i + 1}`, `neptun-wireless-sensor-${i + 1}`));
          this.wirelessSensorServices[i].setCharacteristic(this.platform.Characteristic.Name, `Wireless sensor ${i + 1}`);
        }
      }

      for (let i=0; i<wirelessSensorsCount.wirelessSensorsCount; ++i) {
        const wirelessSensorStatus = await this.neptunSmartModbus.fetchWirelessSensorStatus(i);
        this.platform.log.debug('Updating with', i+1, 'wirelessSensorsStatus:', wirelessSensorStatus);
        this.wirelessSensorServices[i].updateCharacteristic(c.LeakDetected, wirelessSensorStatus.alarmDetected ?
          c.LeakDetected.LEAK_DETECTED : c.LeakDetected.LEAK_NOT_DETECTED);
        this.wirelessSensorServices[i].updateCharacteristic(c.StatusFault, wirelessSensorStatus.connectionLost ?
          c.StatusFault.GENERAL_FAULT : c.StatusFault.NO_FAULT);
        this.wirelessSensorServices[i].updateCharacteristic(c.StatusLowBattery, wirelessSensorStatus.batteryLevel < 20 ?
          c.StatusLowBattery.BATTERY_LEVEL_LOW : c.StatusLowBattery.BATTERY_LEVEL_NORMAL);
      }
    } catch (e) {
      if(e instanceof Error) {
        this.platform.log.error('Could not update config and sensors.', e.message);
      }
    }
  }

  private async handleConfigWrite(writer: ConfigWriter, value: Config, name: string, callback: CharacteristicSetCallback):
    Promise<Config | null> {

    this.platform.log.debug(name, 'Updating to to:', value);

    return writer.call(this.neptunSmartModbus, value)
      .then((result) => {
        this.platform.log.debug(name, 'Update ok. Wrote:', result);
        callback(null);
        return result;
      })
      .catch((error) => {
        this.platform.log.debug(name, 'Update failed. Error:', error.message);
        callback(error);
        return null;
      });
  }

  async handleFaucetFirstGroupActiveSet(value) {
    this.currentConfig.valveOpenFirstGroup = !!value;
    if (!this.isGroupsEnabled) {
      this.currentConfig.valveOpenSecondGroup = !!value;
    }
    this.platform.log.info('Switching valves first group to', value ? 'ACTIVE' : 'INACTIVE');
    await this.handleConfigWrite(this.neptunSmartModbus.writeConfigRegister, this.currentConfig, 'Valve group 1 active', () => {
      this.platform.log.debug('Status for valve group 1 set to', this.currentConfig.valveOpenFirstGroup);
    });
  }

  async handleFaucetSecondGroupActiveSet(value) {
    if (!this.isGroupsEnabled) {
      this.currentConfig.valveOpenFirstGroup = !!value;
    }
    this.currentConfig.valveOpenSecondGroup = !!value;
    this.platform.log.info('Switching valves second group to', value ? 'ACTIVE' : 'INACTIVE');
    await this.handleConfigWrite(this.neptunSmartModbus.writeConfigRegister, this.currentConfig, 'Valve group 2 active', () => {
      this.platform.log.debug('Status for valve group 2 set to', this.currentConfig.valveOpenSecondGroup);
    });
  }
}
